src/ProtoRoute/Message.hs

```diff
@@ -133,24 +133,31 @@ thereExistsMF (MessageField name _) = unFieldName name /= ""
 validateMessage :: MessageSchema -> Message -> Either ValidationError ()
 validateMessage (MessageSchema sn sFields) (Message mn mFields) = do
     verify (mn /= sn) incorrectMsgNames
-    let nameFieldMap = groupedFields
-    let results = DMS.elems nameFieldMap
-    if null (catThese results)
-    then Left invalid
-    else Right ()
+    verify (null $ catThese $ M.elems groupedFields) invalid
   where
-    groupedFields :: DMS.Map FieldName (These SchemaField MessageField)
-    groupedFields = DMS.fromList $ zip fieldsNames theseFields
-
-    fieldsNames =
-        map schemaFieldName sFields `union` map messageFieldName mFields
-    theseFields = zipWith (curry toThese) sFields mFields
-    toThese :: (SchemaField, MessageField) -> These SchemaField MessageField
-    toThese (sf, mf) =
-      if thereExistsSF sf && thereExistsMF mf
-      then These sf mf
-      else if | thereExistsSF sf && not (thereExistsMF mf) -> This sf
-              | not (thereExistsSF sf) && thereExistsMF mf -> That mf
+    groupedFields :: M.Map FieldName (These SchemaField MessageField)
+    groupedFields = unionThese schemaMap msgMap
+
+    unionThese :: Ord k => M.Map k a -> M.Map k b -> M.Map k (These a b)
+    unionThese mapA mapB = sortThemOut (M.assocs mapA) (M.assocs mapB)
+
+    fieldNames :: [FieldName]
+    fieldNames = map schemaFieldName sFields `union` map messageFieldName mFields
+
+    schemaMap :: M.Map FieldName SchemaField
+    schemaMap = M.fromList $ zip fieldNames sFields
+
+    msgMap :: M.Map FieldName MessageField
+    msgMap = M.fromList $ zip fieldNames mFields
+
+    sortThemOut :: Ord k => [(k,a)] -> [(k,b)] -> M.Map k (These a b)
+    sortThemOut [] [] = M.empty
+    sortThemOut ((fn,sf):rest) [] = M.insert fn (This sf) $ sortThemOut rest []
+    sortThemOut [] ((fn,mf):rest) = M.insert fn (That mf) $ sortThemOut [] rest
+    sortThemOut ((sFN,sf):sfs) ((mFN, mf):pairs) =
```

Comment:
The implementation doesn't work. You're still relying on the order of field names.

Instead you can use standard function from `containers` package to implement `unionThese`:

* https://hackage.haskell.org/package/containers-0.6.0.1/docs/Data-Map-Merge-Strict.html#v:merge